pkg/auth/oauth/external/keycloak/keycloak.go

```diff
@@ -0,0 +1,132 @@
+package keycloak
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"strings"
+
+	"github.com/RangelReale/osincli"
+	"github.com/golang/glog"
+
+	authapi "github.com/openshift/origin/pkg/auth/api"
+	"github.com/openshift/origin/pkg/auth/oauth/external"
+)
+
+type provider struct {
+	Realm          string `json:"realm"`
+	RealmPublicKey string `json:"realm-public-key"`
+	ClientID       string `json:"resource"`
+	Credentials    struct {
+		Secret string `json:"secret"`
+	} `json:"credentials"`
+	AuthServerURL string `json:"auth-server-url"`
+}
+
+type keycloakUser struct {
+	id         string
+	username   string
+	email      string
+	firstName  string
+	lastName   string
+	realmRoles string
+}
+
+func NewProvider(keycloakClientConfigFile string) external.Provider {
+	keycloakClientConfigBytes, err := ioutil.ReadFile(keycloakClientConfigFile)
+	if err != nil {
+		glog.Errorf("Error loading Keycloak config: %s", err)
```

Comment:
if an error is a possibility, return multiple values: the provider, and the error. don't make a `nil` return value signal an error.